geospaas_harvesting/crawlers.py

```diff
@@ -377,3 +379,63 @@ def _get_resources_urls(self, xml):
             self._urls.append(entry['link'])
 
         return bool(entries)
+
+
+class FTPCrawler(WebDirectoryCrawler):
+    """
+    Crawler which returns the search results of an FTP, given the URL and search
+    terms
+    """
+    LOGGER = logging.getLogger(__name__ + '.FTPCrawler')
+
+    def __init__(self, root_url, username=None, password=None, fileformat=None,):
+        self.root_url = root_url
+        self.password = password
+        self.username = username
+        self.FILES_SUFFIXES = fileformat
+        self.set_initial_state()
+
+    def set_initial_state(self):
+        """
+        The `_urls` attribute contains URLs to the resources which will be returned by the crawler.
+        The `_to_process` attribute contains URLs to pages which need to be searched for resources.
+        """
+        self._urls = []
+        if not self.root_url.startswith('ftp://'):
+            raise ValueError("root url must start with 'ftp://' in the configuration file")
+        # giving the address that is declared in the config file to "_to_process" attribute
+        self._to_process = [urlparse(self.root_url).path]
+        ftp_domain_name = urlparse(self.root_url).netloc
+        self.ftp = ftplib.FTP(ftp_domain_name, user=self.username, passwd=self.password)
+
+    def _explore_page(self, folder_url):
+        """Get all relevant links from a page and feeds the _urls and _to_process attributes"""
+        self.LOGGER.info("Looking for FTP resources in '%s'...", self.ftp.host+folder_url)
+        try:
+            login_info = self.ftp.login(self.username, self.password)
+        except ftplib.error_perm as e:
+            # these two cases are in the mentioned FTP servers that deals with "login once again"
+            if not (e.args[0].startswith('503') or e.args[0].startswith('230')):
+                raise
+        self.ftp.cwd(folder_url)
+        # for some ftp, pwd is not working properly(returns '')!!! then it has to be set manually
+        # for the first time with the help of python 'or'
+        current_location = self.ftp.pwd() or self.ftp.pwd()+folder_url
+        # searching through all subdirectory to check whether they are folders or files
+        for name in self.ftp.nlst():
```

Comment:
@aperrin66 
`nlst` only gives the names of the folder of the current folder not the full path. It means it returns the last part of the path at each time it has been called. So it is impossible to do it with the same code!

Reply:
Of course you can:
- even with the `nlst` behavior you describe, you just need to keep track of the folder you are in, and store full paths in the `_to_process` stack
- if you give `nlst` an absolute path (beginning with "/"), it will return absolute paths, which makes things much easier.

Reply:
```python
        # searching through all subdirectory to check whether they are folders or files
        for name in self.ftp.nlst(folder_url):
            try:
                self.ftp.cwd(name)
            except ftplib.error_perm as e:
                # if can not cd into new name then add that name to the "self._urls" in the case of
                # having specified endings that are shown in "self.FILES_SUFFIXES"
                if name.endswith(self.FILES_SUFFIXES):
                    ftp_domain_name =self.ftp.host
                    self._urls.append(f"ftp://{ftp_domain_name}{name}")
            else:
                # if successfully cd into new name then add
                # it to "self._to_process" and then come back to original address
                if name not in self._to_process:
                    self._to_process.append(name)
```